hal/spi_async_api.h

```diff
@@ -0,0 +1,182 @@
+/** @{*/
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2018 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef MBED_SPI_API_ASYNC_H
+#define MBED_SPI_API_ASYNC_H
+
+#include "hal/spi_api.h"
+
+#if DEVICE_SPI
+#if DEVICE_SPI_ASYNCH
+
+/** \defgroup hal_spi_async SPI: Asynchronous API.
+ *
+ * This sub API complete the set of functions provided by the SPI API to enable asynchronous access
+ * to the peripheral.
+ *
+ * - Schedule/enqueue a transaction.
+ * - eventually cancel/abort the transaction.
+ * - release the transaction (before or after its completion).
+ *
+ * # Defined behaviour
+ * - `spi_async_transfer_new()` returns NULL if any of `spi_t *obj` or `spi_transfer_args_t *args` is NULL.
+ * - `spi_async_transfer_new()` schedules a transfer using the given parameters.
+ *   All parameters and the embedded references must stay "alive" until completion of the operation.
+ * - `spi_async_transfer_new()` returns a reference counted handle on the scheduled operation.
+ * - `spi_async_transfer_free()` does nothing if passed a NULL pointer.
+ * - `spi_async_transfer_free()` notifies the lowlevel implementation that this reference is no longer
+ *   owned used in the client application (upper/layer code).
+ * - `spi_async_transfer_free()` does **NOT** cancel nor abort a transaction if called before completion.
+ * - `spi_async_transfer_abort()` notifies the lowlevel implementation that the given transaction must be
+ *   cancelled (if not already started) or aborted (if currently running).
+ * - When the operation completes (normally or because of abortion or error) the callback is invoked
+ *   with the provided context and a "reason" describing what triggered the completion.
+ *   This call might be running in an interrupt context and thus all contrainsts applying to ISR
+ *   handler applies to this callback.
+ * - `spi_async_transfer_abort()` does nothing if called on an already completed transaction.
+ * - `spi_async_transfer_abort()` may not wait for the transaction to be cancelled/aborted and returns
+ *   immediatly.
+ * - `spi_free()` cancels and aborts all transactions enqueued for this spi_t.
+ * - the callback passed in spi_async_transfer_new is called only once.
+ *
+ * # Undefined behaviour
+ * - Calling `spi_async_free_handle()` more than one time on a `spi_async_tranfer_t`.
+ * - Calling `spi_async_abort()` after calling `spi_async_free_handle()`.
+ *
+ * # What this API does not cover
```

Comment:
This should be merged with undefined behavior.

Reply:
This async API is still under heavy work. It will be moved indeed.